src/combinatorics/simple_graph/prod.lean

```diff
@@ -173,30 +173,35 @@ by { haveI := (nonempty_prod.1 h.nonempty).1, haveI := (nonempty_prod.1 h.nonemp
 @[simp] lemma box_prod_connected : (G □ H).connected ↔ G.connected ∧ H.connected :=
 ⟨λ h, ⟨h.of_box_prod_left, h.of_box_prod_right⟩, λ h, h.1.box_prod h.2⟩
 
-instance [decidable_eq α] [decidable_eq β] (x : α × β)
+instance box_prod_fintype_neighbor_set (x : α × β)
   [fintype (G.neighbor_set x.1)] [fintype (H.neighbor_set x.2)] :
   fintype ((G □ H).neighbor_set x) :=
-begin
-  rw box_prod_neighbor_set,
-  apply_instance,
-end
+fintype.of_equiv
+  ((G.neighbor_finset x.1 ×ˢ {x.2}).disj_union ({x.1} ×ˢ H.neighbor_finset x.2)
+      $ λ y hG hH, begin
+        simp_rw [finset.mem_product, finset.mem_singleton,
+          mem_neighbor_finset] at hG hH,
+        exact hG.1.ne hH.1.symm,
+      end)
+  ((equiv.refl _).subtype_equiv $ λ y, begin
+    simp_rw [finset.mem_disj_union, finset.mem_product, finset.mem_singleton,
+          mem_neighbor_finset, mem_neighbor_set, equiv.refl_apply, box_prod_adj],
+    simp only [eq_comm, and_comm],
+  end)
+
+lemma box_prod_neighbor_finset (x : α × β)
+  [fintype (G.neighbor_set x.1)] [fintype (H.neighbor_set x.2)] :
+  (G □ H).neighbor_finset x =
+    (G.neighbor_finset x.1 ×ˢ {x.2}).disj_union ({x.1} ×ˢ H.neighbor_finset x.2)
+      (box_prod_fintype_neighbor_set._proof_1 x) :=
```

Comment:
Is it good style to reference `box_prod_fintype_neighbor_set._proof_1`? I'd prefer if we have a separate lemma for it.

Reply:
This is what `@[simps]` would generate (with a worse name), so I think it's probably fine.

I plan to extract this into a separate lemma in #16434 once I can actually use `disjoint` to state it. Right now we're in this awkward mess where using `disj_union` requires a different notion of disjoint to everything else.